### ISSUE TITLE
require-await: triggers with no `await` statement, but returns `Promise`

### ISSUE THREAD

**Issue by nektro** [eslint/eslint#13697]
**Tell us about your environment**

* **ESLint Version:** v6.8.0
* **Node Version:** v14.10.1
* **npm Version:** v6.14.8
* **pnpm Version:** v5.5.4

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
default

**Please show your full configuration:**

https://github.com/nektro/mantle/blob/v49-a2d6e3d/.eslintrc

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

https://github.com/nektro/mantle/blob/v49-a2d6e3d/www/chat/js/api/user.js#L21-L25

```js
async getRoles() {
    return Promise.all(this.roles.map((v) => api.M.roles.get(v))).then((l) => {
        return l.sort((a,b) => a.position - b.position);
    });
}
```

```bash
./node_modules/.bin/eslint ./www/chat/js/
```

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/home/meghan/dev/golang/mantle/www/chat/js/api/user.js
  21:5  error  Async method 'getRoles' has no 'await' expression  require-await
```

**Are you willing to submit a pull request to fix this bug?**
Unfamiliar with eslint code and do not know how one would go about fixing this.

**Comment by ljharb** [comment]
Although this rule is one I think is harmful and shouldn't exist or ever be used, it's erroring correctly in this case because you're using `return Promise.all` instead of `return await Promise.all`.

In other words, the rule is "require await", not "ensure it returns a promise".

**Comment by nektro** [comment]
I see. hmmm, yeah because I'll typically use `async` in this way to show that the function itself can be `await`ed and not necessarily uses `await` itself. So in that case, I should probably turn `require-await` off.

As an aside, are there any other rules that check for `async` and require either `await` OR a `Promise`?

**Comment by mdjermanovic** [comment]
> I see. hmmm, yeah because I'll typically use `async` in this way to show that the function itself can be `await`ed and not necessarily uses `await` itself. So in that case, I should probably turn `require-await` off.

Yes, that seems like a good reason to not use this rule.

> As an aside, are there any other rules that check for `async` and require either `await` OR a `Promise`?

Not in the core rules, and I'm not sure if there's a plugin rule that checks this in JS code.

For TypeScript code, there is [`@typescript-eslint/require-await`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/require-await.md), which works exactly like that.

**Comment by ljharb** [comment]
It's impossible to check "returns a promise" without a fully typed codebase and a type system, which JS itself does not offer.

**Comment by nektro** [comment]
Is this not what jshint doc comments are for?

```js
/** @returns {Promise<api.Role[]>} */
```

**Comment by ljharb** [comment]
@nektro without a typechecker, that's just a comment, which has no semantics in javascript.

**Comment by nektro** [comment]
I'm aware, but does `eslint` not read these, or does it only read the actual code?

**Comment by ljharb** [comment]
I would not expect any builtin eslint rule to treat a nonstandard comment format specially.

Either way, jsdoc comments wouldn't tell you what a function from another file returns, without cross-file linting, which virtually no eslint rules in the entire ecosystem do (outside of eslint-plugin-import)

**Comment by nektro** [comment]
Good to know. Regardless, the name of this issue has been solved. Thanks for the help y'all :)
